tests/units/Services/Table/BuilderTest.php

```diff
@@ -30,235 +30,288 @@ public function setUp(): void
+            'meta' => ['sortable' => true, 'sort' => 'ASC', 'nullLast' => true],
+        ];
+
+        $this->params['meta']['sort'] = true;
+
+        $response = $this->requestResponse();
+
+        $this->assertEquals(
+            $this->secondModel->name,
+            $response->get('data')->first()->get('name')
+        );
     }
 
     /** @test */
     public function can_get_data_with_cache()
     {
         Cache::shouldReceive('get')
-            ->andReturn(-10)
+            ->andReturn(12)
             ->shouldReceive('has')
             ->andReturn(true);
 
-        $response = $this->requestResponse([
-            'cache' => true
-        ]);
+        $this->params['cache'] = true;
 
-        $this->assertEquals(-10, $response["count"]);
+        $response = $this->requestResponse();
+
+        $this->assertEquals(12, $response->get('count'));
     }
 
     /** @test */
     public function can_get_data_with_limit()
     {
-        $response = $this->requestResponse([
-            'meta' => [
-                'length' => 0,
-            ]
-        ]);
+        $this->params['meta']['length'] = 0;
 
-        $this->assertCount(0, $response["data"]);
-    }
+        $response = $this->requestResponse();
 
+        $this->assertCount(0, $response['data']);
+    }
 
     /** @test */
     public function can_get_data_with_total()
     {
-        $response = $this->requestResponse([
-            "columns" => [
-                "id" => ['name' => 'id', "data" => "id", "meta" => ['total' => true]]
-            ],
-            'meta' => [
-                'total' => true,
-            ]
-        ]);
+        $this->createTestModel();
+
+        $this->params['columns']['price'] = [
+            'name' => 'price',
+            'data' => 'price',
+            'meta' => ['total' => true],
+        ];
+
+        $this->params['meta']['total'] = true;
+
+        $response = $this->requestResponse();
 
         $this->assertEquals(
-            BuilderTestModel::all()->sum("id"),
-            $response["total"]['id']
+            BuilderTestModel::sum('price'),
+            $response->get('total')->get('price')
         );
     }
 
-    private function requestResponse(array $params = [], $select = 'id as dtRowId,name,is_active,created_at,price', $withFetch = false)
+    public function can_use_full_info_records_limit()
     {
-        $params['columns'] = $params['columns'] ?? [];
-        $params['meta'] = $params['meta'] ?? [];
-        $params['meta']['length'] = $params['meta']['length'] ?? BuilderTestModel::count();
+        $this->createTestModel();
+
+        $this->params['meta']['fullInfoRecordLimit'] = 1;
+        $this->params['meta']['length'] = 1;
+
+        $response = $this->requestResponse();
+
+        $this->assertEquals(
+            BuilderTestModel::sum('price'),
+            $response->get('total')->get('price')
+        );
+    }
+
+    /** @test */
+    public function can_use_full_info_record_limit()
```

Comment:
this test is about fullRecordInfo. i can't understand below statment
```
$this->fullRecordInfo = $this->meta->get('forceInfo')
            || (! $this->fetchMode && (! $this->filters
                || $this->count <= $this->meta->get('fullInfoRecordLimit')));
```
fullRecordInfo is set there and i can't understand this line.
```
! $this->filters
                || $this->count <= $this->meta->get('fullInfoRecordLimit'))
```
filters always is instance of filter and it is true here. is it correct?
and what's the mean that sentences? why when the filters is false we set fullRecordInfo true even $count is greater than fullInfoRecordLimit(i assume forceInfo and fetchMode is false)

Reply:
>$this->fullRecordInfo = $this->meta->get('forceInfo')
            || (! $this->fetchMode && (! $this->filters
                || $this->count <= $this->meta->get('fullInfoRecordLimit')));

that's ok, I don't either, usually, I always need more than 5 mins to wrap my head around it :)) it's probably one of the worst pieces of code that I ever wrote.

`$this->filters` is not an instance of `Filters` but a boolean. `handle()` from `Filters` returns a bool.